# HW3 report

## Embedding factorization

In [3]:
import embedding_factorization

--- Loading and Preparing Dataset ---


Some weights of BertForTokenClassification were not initialized from the model checkpoint at bert-base-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


--- Loading Model and Applying Factorization ---
Original parameter count: 107,726,601
cpu
Factorized parameter count (rank=64): 87,362,569

--- Starting Model Training ---


W1026 23:30:53.994000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/6] _maybe_guard_rel() was called on non-relation expression Eq(s33, s92) | Eq(s92, 1)
W1026 23:30:53.996000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/6] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)
W1026 23:31:26.761000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/7] _maybe_guard_rel() was called on non-relation expression Eq(s33, s92) | Eq(s92, 1)
W1026 23:31:26.763000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/7] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)
W1026 23:32:31.039000 92806 torch/_dynamo/convert_frame.py:1358] [0/8] torch._dynamo hit config.recompile_limit (8)
W1026 23:32:31.039000 92806 torch/_dynamo/convert_frame.py:1358] [0/8]    function: 'forward' (/home/user/Desktop/venv/lib/python3.10/site-packages/accelerate/utils/operations.py:818)
W1026 23:32:31.039000 92806 torch/_dynamo/co

--- Training Finished ---

           EXPERIMENT RESULTS

--- Final Evaluation on Test Set ---


Metrics on the unseen test set:
  - Test loss: 0.2710
  - Test precision: 0.8546
  - Test recall: 0.8886
  - Test f1: 0.8713
  - Test runtime: 3.9018
  - Test samples_per_second: 884.9700
  - Test steps_per_second: 110.7170


## Parameter sharing

### Attention sharing

In [12]:
import sharing_attention

--- Loading and Preparing Dataset ---


Some weights of BertForTokenClassification were not initialized from the model checkpoint at bert-base-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


--- Loading Model and Applying Factorization ---
Original parameter count: 107,726,601
Factorized parameter count (rank=128): 89,267,465
Shared attention parameter count: 63,264,521

--- Starting Model Training ---


/home/user/Desktop/NLP/hw3/src/sharing_attention.py:64: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  model=model,


--- Training Finished ---

           EXPERIMENT RESULTS

--- Final Evaluation on Test Set ---


Metrics on the unseen test set:
  - Test loss: 0.2731
  - Test precision: 0.8119
  - Test recall: 0.8596
  - Test f1: 0.8351
  - Test runtime: 3.5745
  - Test samples_per_second: 965.9960
  - Test steps_per_second: 120.8540


### Sharing attention + FFN

In [2]:
import sharing_all

--- Loading and Preparing Dataset ---


Some weights of BertForTokenClassification were not initialized from the model checkpoint at bert-base-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


--- Loading Model and Applying Factorization ---
Original parameter count: 107,726,601
Factorized parameter count (rank=128): 89,267,465
Shared attention+FFN parameter count: 11,300,873

--- Starting Model Training ---


/home/user/Desktop/venv/lib/python3.10/site-packages/torch/backends/__init__.py:46: UserWarning: Please use the new API settings to control TF32 behavior, such as torch.backends.cudnn.conv.fp32_precision = 'tf32' or torch.backends.cuda.matmul.fp32_precision = 'ieee'. Old settings, e.g, torch.backends.cuda.matmul.allow_tf32 = True, torch.backends.cudnn.allow_tf32 = True, allowTF32CuDNN() and allowTF32CuBLAS() will be deprecated after Pytorch 2.9. Please see https://pytorch.org/docs/main/notes/cuda.html#tensorfloat-32-tf32-on-ampere-and-later-devices (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:80.)
  self.setter(val)


W1026 22:31:10.252000 64701 torch/fx/experimental/symbolic_shapes.py:6833] [0/1] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)
W1026 22:31:37.171000 64701 torch/fx/experimental/symbolic_shapes.py:6833] [0/2] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)
W1026 22:32:19.762000 64701 torch/fx/experimental/symbolic_shapes.py:6833] [0/3] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)
W1026 22:34:10.103000 64701 torch/fx/experimental/symbolic_shapes.py:6833] [0/4] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)


--- Training Finished ---

           EXPERIMENT RESULTS

--- Final Evaluation on Test Set ---


Metrics on the unseen test set:
  - Test loss: 0.2848
  - Test precision: 0.8006
  - Test recall: 0.8481
  - Test f1: 0.8237
  - Test runtime: 3.5312
  - Test samples_per_second: 977.8590
  - Test steps_per_second: 122.3390


In [8]:
sharing_all.model.bert.embeddings.word_embeddings.embedding.weight

Parameter containing:
tensor([[-7.2949e-02,  6.6831e-02, -1.1975e-02,  ..., -8.1796e-05,
          1.0096e-02, -4.1468e-03],
        [-7.3800e-02,  6.4152e-02, -1.1101e-02,  ...,  7.1646e-03,
          6.3050e-03,  4.8796e-04],
        [-7.3510e-02,  6.4133e-02, -1.0057e-02,  ...,  7.8719e-03,
          5.8717e-03, -3.4092e-03],
        ...,
        [-6.2735e-02,  6.3804e-02, -4.6157e-03,  ...,  5.6066e-03,
          1.4495e-02, -2.2848e-02],
        [-6.2270e-02,  6.0139e-02, -4.8962e-03,  ..., -4.3955e-03,
          1.4510e-02, -7.3002e-03],
        [-6.4505e-02,  6.2615e-02, -5.9869e-03,  ...,  9.2165e-04,
          1.2618e-02, -1.0796e-02]], device='cuda:0', requires_grad=True)

## Layer factorization

In [1]:
import layer_factorization

--- Loading and Preparing Dataset ---
--- Loading Model and Applying Factorization ---


Some weights of BertForTokenClassification were not initialized from the model checkpoint at bert-base-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Original parameter count: 107,726,601
Factorized parameter count (rank=64): 87,362,569
Factorized attention parameter count: 78,567,433


/home/user/Desktop/NLP/hw3/src/layer_factorization.py:70: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(
/usr/bin/ld: cannot find -lcufile: No such file or directory
collect2: error: ld returned 1 exit status



--- Starting Model Training ---


--- Training Finished ---

           EXPERIMENT RESULTS

--- Final Evaluation on Test Set ---


Metrics on the unseen test set:
  - Test loss: 0.2991
  - Test precision: 0.6992
  - Test recall: 0.8052
  - Test f1: 0.7485
  - Test runtime: 4.3314
  - Test samples_per_second: 797.2030
  - Test steps_per_second: 99.7370


На практике качетсво получилось сильно хуже, чем я ожидал. Особенно учитывая, что модель потеряла не так много параметров по сравнению с другими методами.

У меня почему-то не получилось факторизовать dense слой, так как модель начинала ооочень медленно учиться. Обучение можно дополнительно ускорить, если сфьюзить перемножение матриц внутри в одну операцию.

## Distillation

## Base distillation

In [2]:
import distillation_base

--- Loading and Preparing Dataset ---
--- Loading Model and Applying Factorization ---


Some weights of BertForTokenClassification were not initialized from the model checkpoint at bert-base-cased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.
/home/user/Desktop/venv/lib/python3.10/site-packages/torch/backends/__init__.py:46: UserWarning: Please use the new API settings to control TF32 behavior, such as torch.backends.cudnn.conv.fp32_precision = 'tf32' or torch.backends.cuda.matmul.fp32_precision = 'ieee'. Old settings, e.g, torch.backends.cuda.matmul.allow_tf32 = True, torch.backends.cudnn.allow_tf32 = True, allowTF32CuDNN() and allowTF32CuBLAS() will be deprecated after Pytorch 2.9. Please see https://pytorch.org/docs/main/notes/cuda.html#tensorfloat-32-tf32-on-ampere-and-later-devices (Triggered internally at /pytorch/aten/src/ATen/Context.cpp:80.)
  self.setter(val)


Teacher parameter count: 107,726,601

--- Starting Teacher Training ---


/usr/bin/ld: cannot find -lcufile: No such file or directory
collect2: error: ld returned 1 exit status


W1026 23:19:18.552000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/1] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)
W1026 23:19:41.789000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/2] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)
W1026 23:20:05.748000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/3] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)
W1026 23:20:50.589000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/4] _maybe_guard_rel() was called on non-relation expression Eq(s33, s92) | Eq(s92, 1)
W1026 23:20:50.591000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/4] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)


--- Training Finished ---
Student model created with 16,277,641 parameters.
Student model configuration: BertConfig {
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 384,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LABEL_4",
    "5": "LABEL_5",
    "6": "LABEL_6",
    "7": "LABEL_7",
    "8": "LABEL_8"
  },
  "initializer_range": 0.02,
  "intermediate_size": 1536,
  "label2id": {
    "LABEL_0": 0,
    "LABEL_1": 1,
    "LABEL_2": 2,
    "LABEL_3": 3,
    "LABEL_4": 4,
    "LABEL_5": 5,
    "LABEL_6": 6,
    "LABEL_7": 7,
    "LABEL_8": 8
  },
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 8,
  "num_hidden_layers": 8,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "transformers_version": "4.45.2",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size":

W1026 23:23:26.360000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/5] _maybe_guard_rel() was called on non-relation expression Eq(s33, s92) | Eq(s92, 1)
W1026 23:23:26.363000 92806 torch/fx/experimental/symbolic_shapes.py:6833] [0/5] _maybe_guard_rel() was called on non-relation expression Eq(s16, 1) | Eq(s27, s16)


--- Training Finished ---

           EXPERIMENT RESULTS

--- Final Evaluation on Test Set ---


Metrics on the unseen test set:
  - Test loss: 25.8285
  - Test precision: 0.6627
  - Test recall: 0.7130
  - Test f1: 0.6869
  - Test runtime: 2.0627
  - Test samples_per_second: 1674.0250
  - Test steps_per_second: 26.1790


### Attention distillation

#### TODO

# Results

### Результаты оптимизации моделей:

| Метод                 | Параметры (млн) | F1 (Test) | Runtime (Test) |
|-----------------------|-----------------|-----------|----------------|
| Embedding Factorization| 87.36           | 0.8713    | 3.90           |
| Attention Sharing     | 63.26           | 0.8351    | 3.57           |
| Attention + FFN Sharing| 11.30           | 0.8237    | 3.53           |
| Layer Factorization   | 78.57           | 0.7485    | 4.33           |
| Teacher (Distillation)| 107.72          | 0.9420    | -              |
| Student (Distillation)| 16.27           | 0.7586    | 2.06           |

*   **Embedding Factorization:** Лучший баланс между размером и качеством (F1 0.8713 при 87.36M параметров).
*   **Attention + FFN Sharing:** Максимальное сокращение параметров (11.30M), но с F1 0.8237.
*   **Layer Factorization:** Наихудшие показатели F1 среди факторизаций (0.7485). Фокус с низкоранговым приближением всех линейных слоев не прошел.
*   **Distillation:** Student модель очень компактная и быстрая (16.27M, 2.06с), но с F1 0.7586 - требуется доработка.